src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -422,473 +388,307 @@ void IODataBatchOperationListener.BatchOperationContentStreamDisposed()
 
         /// <summary>
-        /// Verifies that calling CreateOperationResponseMessage is valid.
+        /// Verifies that calling CreateOperationRequestMessage is valid.
         /// </summary>
         /// <param name="synchronousCall">true if the call is to be synchronous; false otherwise.</param>
-        private void VerifyCanCreateOperationResponseMessage(bool synchronousCall)
-        {
-            this.ValidateWriterReady();
-            this.VerifyCallAllowed(synchronousCall);
-
-            if (!this.rawOutputContext.WritingResponse)
-            {
-                this.ThrowODataException(Strings.ODataBatchWriter_CannotCreateResponseOperationWhenWritingRequest);
-            }
-        }
+        /// <param name="method">The Http method to be used for this request operation.</param>
+        /// <param name="uri">The Uri to be used for this request operation.</param>
+        /// <param name="contentId">The Content-ID value to write in ChangeSet head.</param>
+        protected abstract void VerifyCanCreateOperationRequestMessage(bool synchronousCall, string method, Uri uri,
+            string contentId);
 
         /// <summary>
-        /// Creates an <see cref="ODataBatchOperationResponseMessage"/> for writing an operation of a batch response - implementation of the actual functionality.
+        /// Sets a new writer state; verifies that the transition from the current state into new state is valid.
         /// </summary>
-        /// <param name="contentId">The Content-ID value to write in ChangeSet head.</param>
-        /// <returns>The message that can be used to write the response operation.</returns>
-        private ODataBatchOperationResponseMessage CreateOperationResponseMessageImplementation(string contentId)
-        {
-            this.WritePendingMessageData(true);
-
-            // In responses we don't need to use our batch URL resolver, since there are no cross referencing URLs
-            // so use the URL resolver from the batch message instead.
-            this.CurrentOperationResponseMessage = ODataBatchOperationResponseMessage.CreateWriteMessage(
-                this.rawOutputContext.OutputStream,
-                /*operationListener*/ this,
-                this.payloadUriConverter.BatchMessagePayloadUriConverter,
-                this.container);
-            this.SetState(BatchWriterState.OperationCreated);
-
-            Debug.Assert(this.currentOperationContentId == null, "The Content-ID header is only supported in request messages.");
+        /// <param name="newState">The writer state to transition into.</param>
+        protected abstract void SetState(BatchWriterState newState);
 
-            // write the operation's start boundary string
-            this.WriteStartBoundaryForOperation();
+        /// <summary>
+        /// Validates that the batch writer is ready to process a new write request.
+        /// </summary>
+        protected abstract void ValidateWriterReady();

```

Comment:
>ValidateWriterReady [](start = 32, length = 19)

Rather than expose ValidateWriterReady, for consistency with ODataWriterCore, expose protected a abstract VerifyNotDisposed. Your calls to ValidateWriterReady() should call a private ValdiateWriterReady that first calls VerifyNotDisposed in the derived class, then does the state check that is currently in the ODataMultipartMixedBatchWriter implementation of ValidateWriterReady() (but common to any writer).

Reply:
make senses. Now exposing abstract VerifyNotDisposed in base class, and make ValidateWriterReady in base class.

---
In reply to: [140934964](https://github.com/OData/odata.net/pull/960#discussion_r140934964) [](ancestors = 140934964)